plutus-core/testlib/PlutusIR/Properties/Typecheck.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+
+{- | Typechecking property tests for all PIR compiler passes. Currently we only typecheck the term
+ after a pass.
+
+ There were test failures when we infer the type before a pass and check that the
+ type after a pass is the same. The failures are probably caused by the
+ `PlutusIR.TypeCheck.AllowEscape` typechecker config.
+-}
+module PlutusIR.Properties.Typecheck where
+
+import Control.Monad.Reader
+import PlutusCore.Default
+import PlutusCore.Generators.QuickCheck.Utils
+import PlutusCore.Quote
+import PlutusCore.Rename
+import PlutusCore.TypeCheck qualified as PLC
+import PlutusIR.Compiler
+import PlutusIR.Compiler.Provenance (original)
+import PlutusIR.Core
+import PlutusIR.Generators.QuickCheck.GenerateTerms
+import PlutusIR.TypeCheck
+import PlutusPrelude (($>))
+import Test.QuickCheck (Property, elements)
+import Test.QuickCheck.Arbitrary
+
+instance Arbitrary (BuiltinSemanticsVariant DefaultFun) where
+    arbitrary = elements [DefaultFunSemanticsVariant1, DefaultFunSemanticsVariant2]
+
+-- Convert Either Error () to Either String () to match with the Testable (Either String ())
+-- instance.
+convertToEitherString :: Either (Error DefaultUni DefaultFun ()) ()
+  -> Either String ()
+convertToEitherString = \case
+  Left err -> Left $ show err
+  Right () -> Right ()
+
+-- | Check that a term typechecks after one of the pure passes.
+pure_typecheck_prop ::
+  -- | The pure simplification pass.
+  (Term TyName Name DefaultUni DefaultFun ()
+  -> Term TyName Name DefaultUni DefaultFun ())
+  -> Property
+pure_typecheck_prop pass =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) ->
+    convertToEitherString $ do
+      config <- getDefTypeCheckConfig ()
+      -- the generated term may not have globally unique names
+      renamed <- runQuoteT $ rename tm
+      _ <- runQuoteT $ inferType config (pass renamed)
```

Comment:
You should be running all the `QuoteT` stuff together, otherwise you might generate duplicate names. I think that's safe here, but might not be in general